libs/x-content/src/test/java/org/elasticsearch/xcontent/XContentParserTests.java

```diff
@@ -475,6 +475,41 @@ public void testSubParserObject() throws IOException {
         }
     }
 
+    public void testFlatteningParserObject() throws IOException {
+        String content = """
+            {
+              "parent": {
+                "child1" : 1,
+                "child2": {
+                  "grandChild" : 1
+                },
+                "child3" : 1
+              }
+            }
+            """;
+        XContentParser parser = createParser(JsonXContent.jsonXContent, content);
+        assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
+        assertEquals(XContentParser.Token.FIELD_NAME, parser.nextToken());
+        assertEquals("parent", parser.currentName());
+        assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
+        XContentParser subParser = new FlatteningXContentParser(parser, parser.currentName());
+        assertEquals(XContentParser.Token.FIELD_NAME, subParser.nextToken());
+        assertEquals("parent.child1", subParser.currentName());
+        assertEquals(XContentParser.Token.VALUE_NUMBER, subParser.nextToken());
+        assertEquals(XContentParser.Token.FIELD_NAME, subParser.nextToken());
+        String secondChildName = subParser.currentName();
+        assertEquals("parent.child2", secondChildName);
+        assertEquals(XContentParser.Token.START_OBJECT, subParser.nextToken());
+        assertEquals(XContentParser.Token.FIELD_NAME, subParser.nextToken());
+        assertEquals("grandChild", subParser.currentName());
```

Comment:
can you help me understand why we need to prepend the path only if level is 1? shouldn't this be also `parent.child2.grandChild` ?

Reply:
Every time that we have a dynamically created object mapper and subobjects:false we create a new context with a FlatteningXContentParser. This is explicit when we create `subParser` for field `parent` and is not done (for tests purpouses) for the field `child2`. In the normal code execution since 'child2' is a dynamically created object mapper itself we would have created a sub-subParser passing as parent name `subParser.currentName()` that is `parent.child2` as was done in [this test](https://github.com/elastic/elasticsearch/pull/97972#discussion_r1304248220) and therefore the current name here would have being `parent.child2.grandChild`.
Is for this reason that we only "care" about immediate child, since grandchildren are handled recursively. Otherwise we would have created somewhat of a clone of ContentPath.

Reply:
Got it, thanks for your patient explanation ;)